### PR TITLE
added sidecar.istio.io/rewriteAppProbers annotation

### PR DESCRIPTION
**Pull request by arief-hidayat**
For #13074

* register sidecar.istio.io/rewriteAppProbers annotation
* include this annotation upon checking whether should rewrite app probers (previously we checked only RewriteAppHTTPProbe flag in SidecarInjectionSpec) 
* tests added

### CODE REVIEW

**Comment by arief-hidayat** [comment]
I signed it!

File: pilot/pkg/kube/inject/app_probe_test.go [hunk 1]
```diff
@@ -43,3 +43,23 @@ func TestFindSidecar(t *testing.T) {
...
+	for _, tc := range []struct {
+		name                 string
+		sidecarInjectionSpec SidecarInjectionSpec
+		annotations          map[string]string
+		expected             bool
+	}{
+		{"RewriteAppHTTPProbe-not-set", SidecarInjectionSpec{RewriteAppHTTPProbe: false}, map[string]string{}, false},
+		{"RewriteAppHTTPProbe-set-in-annotations", SidecarInjectionSpec{RewriteAppHTTPProbe: false}, map[string]string{annotationRewriteAppProbers: "true"}, true},
+		{"RewriteAppHTTPProbe-set-in-sidecar-injection-spec", SidecarInjectionSpec{RewriteAppHTTPProbe: true}, map[string]string{}, true},
+		{"RewriteAppHTTPProbe-set-in-sidecar-injection-spec-&-annotations", SidecarInjectionSpec{RewriteAppHTTPProbe: true}, map[string]string{annotationRewriteAppProbers: "true"}, true},
```
Comment (golangcibot) [inline pilot/pkg/kube/inject/app_probe_test.go]: line is 181 characters (from `lll`)

File: pilot/pkg/kube/inject/inject.go [hunk 2]
```diff
@@ -51,8 +51,9 @@ import (
 type annotationValidationFunc func(value string) error
 
 const (
-	annotationPolicy = "sidecar.istio.io/inject"
-	annotationStatus = "sidecar.istio.io/status"
+	annotationPolicy            = "sidecar.istio.io/inject"
+	annotationStatus            = "sidecar.istio.io/status"
+	annotationRewriteAppProbers = "sidecar.istio.io/rewriteAppProbers"
```
Comment (incfly) [inline pilot/pkg/kube/inject/inject.go]: Can you change to `rewriteAppHTTPProbers`? We use that naming throughout helm flags, and our code. Keeping it consistent can avoid confusion for future developers.

k8s app can also have tcp probers, which is not the under the scope of this feature.

File: pilot/pkg/kube/inject/inject.go [hunk 3]
```diff
@@ -64,6 +65,7 @@ var (
 	annotationRegistry = map[string]annotationValidationFunc{
 		annotations.Register(annotationPolicy, "").Name:                                        alwaysValidFunc,
 		annotations.Register(annotationStatus, "").Name:                                        alwaysValidFunc,
+		annotations.Register(annotationRewriteAppProbers, "").Name:                             validateBool,
```
Comment (incfly) [inline pilot/pkg/kube/inject/inject.go]: annotations.Register() second argument is the description of the annotations. I assume @geeknoid intention is to auto generate the document at some point. Previous annotation does not have doc, but let's add for new one.

Some brief sentence about what this annotation is for is good enough.

File: pilot/pkg/kube/inject/app_probe.go [hunk 4]
```diff
@@ -27,6 +27,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/intstr"
 
 	"istio.io/istio/pilot/cmd/pilot-agent/status"
+
```
Comment (incfly) [inline pilot/pkg/kube/inject/app_probe.go]: no need for this blank line?

File: pilot/pkg/kube/inject/app_probe_test.go [hunk 5]
```diff
@@ -43,3 +43,24 @@ func TestFindSidecar(t *testing.T) {
...
 }
+
+func TestShouldRewriteAppProbers(t *testing.T) {
+	for _, tc := range []struct {
+		name                 string
+		sidecarInjectionSpec SidecarInjectionSpec
+		annotations          map[string]string
+		expected             bool
+	}{
+		{"RewriteAppHTTPProbe-not-set", SidecarInjectionSpec{RewriteAppHTTPProbe: false}, map[string]string{}, false},
```
Comment (incfly) [inline pilot/pkg/kube/inject/app_probe_test.go]: can you break struct into several lines for better readability? currently this struct is a bit long and hard to follow.

File: pilot/pkg/kube/inject/app_probe_test.go [hunk 6]
```diff
@@ -43,3 +43,24 @@ func TestFindSidecar(t *testing.T) {
...
+		{"RewriteAppHTTPProbe-set-in-annotations", SidecarInjectionSpec{RewriteAppHTTPProbe: false}, map[string]string{annotationRewriteAppProbers: "true"}, true},
```
Comment (incfly) [inline pilot/pkg/kube/inject/app_probe_test.go]: add another test case where annotationsMap is nil?

File: pilot/pkg/kube/inject/testdata/webhook/TestWebhookInject_http_probe_rewrite_disabled_via_annotation_template.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
+initContainers:
```
Comment (incfly) [inline pilot/pkg/kube/inject/testdata/webhook/TestWebhookInject_http_probe_rewrite_disabled_via_annotation_template.yaml]: add `rewriteAppHTTPProbers: true`? Otherwise I suppose we generate the same patch because the helm flag is not enabled even before this PR.

**Comment by incfly** [comment]
add pilot owners.

**Comment by incfly** [comment]
Another thing if you are interested, is to add this as a negative test into our e2e testing framework.

https://github.com/istio/istio/blob/master/tests/integration/security/healthcheck/mtls_healthcheck_test.go#L82

Basically, using annotations to turn off the feature and then verifies that `aps.GetAppOrFail("healthcheck-disabled-via-annotations", t)` will fail.

Purely optional, I can also do it later.

File: pilot/pkg/kube/inject/app_probe_test.go (hunk 5 above)
Comment (arief-hidayat) [reply pilot/pkg/kube/inject/app_probe_test.go]: changed. please have a look

File: pilot/pkg/kube/inject/app_probe_test.go (hunk 6 above)
Comment (arief-hidayat) [reply pilot/pkg/kube/inject/app_probe_test.go]: added more test cases including nil case

File: pilot/pkg/kube/inject/testdata/webhook/TestWebhookInject_http_probe_rewrite_disabled_via_annotation_template.yaml (hunk 7 above)
Comment (arief-hidayat) [reply pilot/pkg/kube/inject/testdata/webhook/TestWebhookInject_http_probe_rewrite_disabled_via_annotation_template.yaml]: added thanks.

File: pilot/pkg/kube/inject/app_probe.go (hunk 4 above)
Comment (arief-hidayat) [reply pilot/pkg/kube/inject/app_probe.go]: removed

File: pilot/pkg/kube/inject/inject.go (hunk 3 above)
Comment (arief-hidayat) [reply pilot/pkg/kube/inject/inject.go]: added description. please let me know if you have better desc

File: pilot/pkg/kube/inject/inject.go (hunk 2 above)
Comment (arief-hidayat) [reply pilot/pkg/kube/inject/inject.go]: I have renamed to `rewriteAppHTTPProbers`. However, the one in template is `rewriteAppHTTPProbe`. Should we keep both in sync?

**Comment by arief-hidayat** [comment]
Hi @incfly 
Thank you for the review comments. I have made corresponding changes except for e2e test framework. I am going to get familiarised and add the test later.

Thanks

File: pilot/pkg/kube/inject/inject.go [hunk 8]
```diff
@@ -62,8 +63,10 @@ var (
 	}
 
 	annotationRegistry = map[string]annotationValidationFunc{
-		annotations.Register(annotationPolicy, "").Name:                                        alwaysValidFunc,
-		annotations.Register(annotationStatus, "").Name:                                        alwaysValidFunc,
+		annotations.Register(annotationPolicy, "").Name: alwaysValidFunc,
+		annotations.Register(annotationStatus, "").Name: alwaysValidFunc,
+		annotations.Register(annotationRewriteAppHTTPProbers,
+			"Rewrite HTTP readiness and liveness probes to be redirected to istio-proxy sidecar").Name: validateBool,
```
Comment (hzxuzhonghu) [inline pilot/pkg/kube/inject/inject.go]: Will this fail the inject? I would leave it not influence the normal injection, just ignore this annotation.

**Comment by hzxuzhonghu** [comment]
Overall, the code part lgtm.

File: pilot/pkg/kube/inject/inject.go (hunk 8 above)
Comment (arief-hidayat) [reply pilot/pkg/kube/inject/inject.go]: Sorry @hzxuzhonghu , I don't get your point due to my lack understanding of the context.
Which injection will fail?
Do you mean to use `alwaysValidFunc` instead of `validateBool` ?

Comment (hzxuzhonghu) [reply pilot/pkg/kube/inject/inject.go]: yes, I mean if we set for example $annotationRewriteAppHTTPProbers: invalid-value, it will fail

Comment (arief-hidayat) [reply pilot/pkg/kube/inject/inject.go]: I think it's better to have it validated.
anyway, I will change to `alwaysValidFunc`
thanks
